content/en/tutorials/reproducible-machine-learning-cloud-pods/index.md

```diff
@@ -0,0 +1,230 @@
+
+```python
+def load_digits(*, n_class=10, return_X_y=False, as_frame=False):
+    # download files from S3
+    s3_client = boto3.client("s3")
+    s3_client.download_file(Bucket="pods-test", Key="digits.csv.gz", Filename="digits.csv.gz")
+
+    data = numpy.loadtxt('digits.csv.gz', delimiter=',')
+    target = data[:, -1].astype(numpy.int, copy=False)
+    flat_data = data[:, :-1]
+    images = flat_data.view()
+    images.shape = (-1, 8, 8)
+
+    if n_class < 10:
+        idx = target < n_class
+        flat_data, target = flat_data[idx], target[idx]
+        images = images[idx]
+
+    feature_names = ['pixel_{}_{}'.format(row_idx, col_idx)
+                     for row_idx in range(8)
+                     for col_idx in range(8)]
+
+    frame = None
+    target_columns = ['target', ]
+    if as_frame:
+        frame, flat_data, target = datasets._convert_data_dataframe(
+            "load_digits", flat_data, target, feature_names, target_columns)
+
+    if return_X_y:
+        return flat_data, target
+
+    return Bunch(data=flat_data,
+                 target=target,
+                 frame=frame,
+                 feature_names=feature_names,
+                 target_names=numpy.arange(10),
+                 images=images)
+```
+
+The above code uses the `boto3` library to download the data file from an S3 bucket. The file is then loaded into a NumPy array using the `numpy.loadtxt` function, and the target values (i.e. the labels corresponding to each image) are extracted from the last column of the array. The images are then reshaped into 2-dimensional arrays, and the function has been configured to return only a subset of the available classes by filtering the target values. Finally, the function returns an object containing the data, target values, and metadata.
+
+Let us now define a `handler` function that would be executed by the Lambda every time a trigger event occurs. In this case, we would like to use the above function to load the dataset and train a model using the Support Vector Machine (SVM) algorithm. After a train-test split, we will predict the value of the digit on the test subset:
+
+```python
+def handler(event, context):
+
+    digits = load_digits()
+
+    # flatten the images
+    n_samples = len(digits.images)
+    data = digits.images.reshape((n_samples, -1))
+
+    # Create a classifier: a support vector classifier
+    clf = svm.SVC(gamma=0.001)
+
+    # Split data into 50% train and 50% test subsets
+    X_train, X_test, y_train, y_test = train_test_split(
+        data, digits.target, test_size=0.5, shuffle=False
+    )
+
+    # Learn the digits on the train subset
```

Comment:
Thinking a bit more about our sample, I think it would make sense to split up the training and inference parts. We only created this initial version as a quick first version for the blog post, but to make it more realistic and to highlight the sharing / collaboration aspect, we could split up the sample into two parts:

* `training` Lambda: loads the digits data set, trains the model via `fit(..)`, stores the pickled model object to S3
* `inference` Lambda: loads the pickled model file from S3, runs a single inference by calling `predict(..)` on the model

This way the sample would be more illustrative of the state sharing feature, as not only the training files (digits) but also the actual trained model would become part of the Cloud Pod. (In the current implementation, the model is retrained on each invocation of the Lambda)

Thoughts @HarshCasper @giograno ? Depending on our planned timeline, we could also add this as a follow-up iteration (to get a first version out quickly). Also, happy to help out with the implementation of this..

Reply:
Good input @whummer 🙏  I can take care of it next week, if there is no particular rush to go out.

Reply:
@whummer I reworked the sample as you suggested in https://github.com/localstack/presentations/pull/1. What do you think about it?